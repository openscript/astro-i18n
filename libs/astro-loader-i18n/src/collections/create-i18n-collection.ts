import { I18nCollection } from "../schemas/i18n-loader-schema";

type Options = {
  locales: string[];
  routePattern: string;
  basePath?: string;
};

/**
 * Creates an internationalization (i18n) collection based on the provided options.
 *
 * @param options - Configuration options for the i18n collection.
 * @returns An array of objects representing the i18n collection for each locale.
 */
export function createI18nCollection(options: Options): I18nCollection {
  const { locales, routePattern, basePath } = options;

  return locales.map((locale) => ({ data: { locale: locale, translationId: routePattern, contentPath: "", basePath: basePath || "" } }));
}
