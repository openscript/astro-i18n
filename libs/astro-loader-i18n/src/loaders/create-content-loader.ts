import { createContentPath, createTranslationId, getAllUniqueKeys, pruneLocales } from "astro-utils-i18n";
import { glob, Loader, LoaderContext } from "astro/loaders";

const UNDETERMINED_LOCALE = "und";
type GlobOptions = Parameters<typeof glob>[0];

export function createContentLoader(loader: Loader, base?: GlobOptions["base"]) {
  return async (context: LoaderContext) => {
    if (!context.config.i18n) throw new Error("i18n configuration is missing in your astro config");

    const { locales } = context.config.i18n;
    const localeCodes = locales.flatMap((locale) => (typeof locale === "string" ? locale : locale.codes));

    const parseData = context.parseData;
    const parseDataProxy: typeof parseData = (props) => {
      if (!props.filePath) return parseData(props);
      const locale = UNDETERMINED_LOCALE;
      const translationId = createTranslationId(props.filePath);
      const contentPath = createContentPath(props.filePath, base);
      const basePath = context.config.base;
      return parseData({ ...props, data: { ...props.data, locale, translationId, contentPath, basePath } });
    };
    context.parseData = parseDataProxy;

    await loader.load(context);

    const entries = context.store.entries();
    context.store.clear();

    entries.forEach(([, entry]) => {
      const entryLocales = Array.from(getAllUniqueKeys(entry.data)).filter((key) => localeCodes.includes(key));
      if (entryLocales.length === 0) {
        context.store.set(entry);
      } else {
        entryLocales.forEach((locale) => {
          const entryData = pruneLocales(entry.data, entryLocales, locale);
          context.store.set({ ...entry, id: `${entry.id}/${locale}`, data: { ...entryData, locale } });
        });
      }
    });
  };
}
