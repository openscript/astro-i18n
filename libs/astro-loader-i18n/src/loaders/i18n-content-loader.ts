import { glob, Loader } from "astro/loaders";
import { createContentLoader } from "./create-content-loader";

type GlobOptions = Parameters<typeof glob>[0];

/**
 * A loader function for handling internationalization (i18n) content in an Astro project.
 * This loader processes files matching the specified glob pattern, associates them with locales,
 * and augments their data with i18n-specific metadata such as locale, translation ID, and content path.
 *
 * @param options - Configuration options for the glob pattern to match files.
 * @returns A loader object with a custom `load` method for processing i18n content.
 *
 * @throws Will throw an error if the `i18n` configuration is missing in the Astro config.
 */
export function i18nContentLoader(options: GlobOptions): Loader {
  const globLoader = glob(options);
  return {
    name: "i18n-content-loader",
    load: createContentLoader(globLoader, options.base),
  };
}
