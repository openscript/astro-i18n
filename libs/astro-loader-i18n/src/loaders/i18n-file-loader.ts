import { file, Loader, LoaderContext } from "astro/loaders";
import { pruneLocales, getAllUniqueKeys, createContentPath, createTranslationId } from "astro-utils-i18n";

const UNDETERMINED_LOCALE = "und";

type FileOptions = Parameters<typeof file>;

/**
 * A loader function for handling internationalization (i18n) content in an Astro project.
 * This loader processes a single file and augments their data with i18n-specific metadata
 * such as locale, translation ID, and content path.
 *
 * @param fileName - The name of the file to be processed.
 * @param options - Configuration options for matching the file.
 * @returns A loader object with a custom `load` method for processing i18n content.
 *
 * @throws Will throw an error if the `i18n` configuration is missing in the Astro config.
 */
export function i18nFileLoader(fileName: FileOptions[0], options?: FileOptions[1]): Loader {
  const fileLoader = file(fileName, options);
  return {
    name: "i18n-file-loader",
    load: async (context: LoaderContext) => {
      if (!context.config.i18n) throw new Error("i18n configuration is missing in your astro config");

      const { locales } = context.config.i18n;
      const localeCodes = locales.flatMap((locale) => (typeof locale === "string" ? locale : locale.codes));

      const parseData = context.parseData;
      const parseDataProxy: typeof parseData = (props) => {
        if (!props.filePath) return parseData(props);
        const locale = UNDETERMINED_LOCALE;
        const translationId = createTranslationId(props.filePath);
        const contentPath = createContentPath(props.filePath);
        const basePath = context.config.base;
        return parseData({ ...props, data: { ...props.data, locale, translationId, contentPath, basePath } });
      };
      context.parseData = parseDataProxy;

      await fileLoader.load(context);

      const entries = context.store.entries();
      context.store.clear();

      entries.forEach(([, entry]) => {
        const entryLocales = Array.from(getAllUniqueKeys(entry.data)).filter((key) => localeCodes.includes(key));
        if (entryLocales.length === 0) {
          context.store.set(entry);
        } else {
          entryLocales.forEach((locale) => {
            const entryData = pruneLocales(entry.data, entryLocales, locale);
            context.store.set({ ...entry, id: `${entry.id}/${locale}`, data: { ...entryData, locale } });
          });
        }
      });
    },
  };
}
