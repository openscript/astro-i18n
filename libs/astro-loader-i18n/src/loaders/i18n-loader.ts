import { glob, Loader, LoaderContext } from "astro/loaders";
import { createContentPath, createTranslationId, parseLocale } from "astro-utils-i18n";

type GlobOptions = Parameters<typeof glob>[0];

/**
 * Creates a custom i18n loader for Astro projects.
 *
 * @param options - Configuration options for the glob loader.
 * @returns A loader that integrates i18n functionality into the Astro build process.
 *
 * @throws If the `i18n` configuration is missing in the Astro config.
 */
export function i18nLoader(options: GlobOptions): Loader {
  const globLoader = glob(options);
  return {
    name: "i18n-loader",
    load: async (context: LoaderContext) => {
      if (!context.config.i18n) throw new Error("i18n configuration is missing in your astro config");

      const { locales, defaultLocale } = context.config.i18n;
      const localeCodes = locales.flatMap((locale) => (typeof locale === "string" ? locale : locale.codes));

      const parseData = context.parseData;
      const parseDataProxy: typeof parseData = (props) => {
        if (!props.filePath) return parseData(props);
        const locale = parseLocale(props.filePath, localeCodes, defaultLocale);
        const translationId = createTranslationId(props.filePath, locale);
        const contentPath = createContentPath(props.filePath, options.base, locale);
        const basePath = context.config.base;
        return parseData({ ...props, data: { ...props.data, locale, translationId, contentPath, basePath } });
      };
      context.parseData = parseDataProxy;

      await globLoader.load(context);
    },
  };
}
