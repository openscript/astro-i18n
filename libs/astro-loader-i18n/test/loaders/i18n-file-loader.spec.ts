import { beforeEach, describe, expect, it, vi } from "vitest";
import { LoaderContext } from "astro/loaders";
import { createLoaderContext } from "../__mocks__/loader-context";
import { contentFixture } from "../__fixtures__/collections";
import { i18nFileLoader } from "../../src/loaders/i18n-file-loader";

vi.mock("astro/loaders", () => {
  return {
    file: () => {
      return {
        load: vi.fn().mockImplementation(async (context: LoaderContext) => {
          contentFixture.forEach(async (entry) => {
            context.store.set({ ...entry, data: await context.parseData(entry) });
          });
        }),
      };
    },
  };
});

describe("i18nFileLoader", () => {
  let context: LoaderContext;

  beforeEach(() => {
    context = createLoaderContext();
  });

  it("should put common translation id and locale in data", async () => {
    const loader = i18nFileLoader("/content/gallery/space.yml");
    await loader.load(context);

    const entries = context.store.entries();
    expect(entries).toMatchSnapshot();
  });

  it("should throw error if i18n config is missing", async () => {
    const loader = i18nFileLoader("/content/gallery/space.yml");
    context.config.i18n = undefined;

    await expect(loader.load(context)).rejects.toThrowErrorMatchingSnapshot();
  });
});
