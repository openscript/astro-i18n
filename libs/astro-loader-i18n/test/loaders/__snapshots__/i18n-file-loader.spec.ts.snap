// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`i18nFileLoader > should handle input without i18n 1`] = `
[
  [
    "omni",
    {
      "data": {
        "basePath": "/",
        "contentPath": "src/content/gallery",
        "cover": "./animals1.jpg",
        "images": [
          {
            "src": "./animals1.jpg",
          },
          {
            "src": "./animals2.jpg",
          },
          {
            "src": "./animals3.jpg",
          },
          {
            "src": "./animals4.jpg",
          },
          {
            "src": "./animals5.jpg",
          },
        ],
        "locale": "und",
        "title": "Omni",
        "translationId": "src/content/gallery/omni.yml",
      },
      "filePath": "src/content/gallery/omni.yml",
      "id": "omni",
    },
  ],
]
`;

exports[`i18nFileLoader > should put common translation id and locale in data 1`] = `
[
  [
    "space/de-CH",
    {
      "data": {
        "basePath": "/",
        "contentPath": "src/content/gallery",
        "cover": "./space1.jpg",
        "images": [
          {
            "src": "./space1.jpg",
            "title": "Weltraum1",
          },
          {
            "src": "./space2.jpg",
            "title": "Weltraum2",
          },
          {
            "src": "./space3.jpg",
            "title": "Weltraum3",
          },
          {
            "src": "./space4.jpg",
            "title": "Weltraum4",
          },
          {
            "src": "./space5.jpg",
            "title": "Weltraum5",
          },
        ],
        "locale": "de-CH",
        "title": "Weltraum",
        "translationId": "src/content/gallery/space.yml",
      },
      "filePath": "src/content/gallery/space.yml",
      "id": "space/de-CH",
    },
  ],
  [
    "space/zh-CN",
    {
      "data": {
        "basePath": "/",
        "contentPath": "src/content/gallery",
        "cover": "./space1.jpg",
        "images": [
          {
            "src": "./space1.jpg",
            "title": "Space1",
          },
          {
            "src": "./space2.jpg",
            "title": "Space2",
          },
          {
            "src": "./space3.jpg",
            "title": "Space3",
          },
          {
            "src": "./space4.jpg",
            "title": "Space4",
          },
          {
            "src": "./space5.jpg",
            "title": "Space5",
          },
        ],
        "locale": "zh-CN",
        "title": "太空",
        "translationId": "src/content/gallery/space.yml",
      },
      "filePath": "src/content/gallery/space.yml",
      "id": "space/zh-CN",
    },
  ],
  [
    "gallery/de-CH",
    {
      "data": {
        "basePath": "/",
        "contentPath": "src/content/gallery",
        "items": [
          {
            "label": "Robin improvisiert beim Arbeiten",
            "photo": "./team/robin.jpg",
          },
          {
            "label": "Robin präsentiert",
            "photo": "./team/robin-presents.jpeg",
          },
          {
            "label": "Unser Labor",
            "photo": "./team/our-lab.jpg",
          },
          {
            "label": "Team-Retreat 2024",
            "photo": "./team/retraite-2024.jpg",
          },
        ],
        "locale": "de-CH",
        "translationId": "src/content/gallery/gallery.yml",
      },
      "filePath": "src/content/gallery/gallery.yml",
      "id": "gallery/de-CH",
    },
  ],
  [
    "gallery/zh-CN",
    {
      "data": {
        "basePath": "/",
        "contentPath": "src/content/gallery",
        "items": [
          {
            "label": "Robin即兴工作",
            "photo": "./team/robin.jpg",
          },
          {
            "label": "Robin演示",
            "photo": "./team/robin-presents.jpeg",
          },
          {
            "label": "我们的实验室",
            "photo": "./team/our-lab.jpg",
          },
          {
            "label": "团队静修2024",
            "photo": "./team/retraite-2024.jpg",
          },
        ],
        "locale": "zh-CN",
        "translationId": "src/content/gallery/gallery.yml",
      },
      "filePath": "src/content/gallery/gallery.yml",
      "id": "gallery/zh-CN",
    },
  ],
]
`;

exports[`i18nFileLoader > should throw error if i18n config is missing 1`] = `[Error: i18n configuration is missing in your astro config]`;
